core/src/main/scala/fs2/io/ssh/Client.scala

```diff
@@ -17,85 +17,112 @@
+import org.apache.sshd.common.config.keys.FilePasswordProvider
+import org.apache.sshd.common.keyprovider.FileKeyPairProvider
 
 import scala.{Array, Int, None, Some}
+import scala.collection.JavaConverters._
 
 import java.lang.{String, SuppressWarnings}
 import java.net.{InetAddress, InetSocketAddress}
 
-object Client {
+final class Client[F[_]: Concurrent: ContextShift] private (client: SshClient) {
+  import MinaFuture._
+
+  private val F = Concurrent[F]
 
-  @SuppressWarnings(Array("org.wartremover.warts.DefaultArguments"))
-  def exec[F[_]: Sync: ContextShift](
-      host: InetSocketAddress,
-      username: String,
-      auth: Auth,
+  @SuppressWarnings(
+    Array(
+      "org.wartremover.warts.DefaultArguments",
+      "org.wartremover.warts.Null",
+      "org.wartremover.warts.ToString"))
+  def exec(
+      cc: ConnectionConfig,
       command: String,
       blocker: Blocker,
       chunkSize: Int = 4096)
-      : Resource[F, Process[F]] =
-    connect[F](host, username, auth, blocker) flatMap { session =>
-      val r = Resource.make(blocker.blockOn(Sync[F].delay(session.exec(command)))) { cmd =>
-        blocker.blockOn(Sync[F].delay(cmd.close()))
+      : Resource[F, Process[F]] = {
+
+    for {
+      session <-
+        Resource.make(
+          fromFuture(F.delay(client.connect(cc.username, cc.host))))(
+          session => fromFuture(F.delay(session.close(false))).void)
+
+      _ <- cc.auth match {
+        case Auth.Password(text) =>
+          Resource.liftF(F.delay(session.addPasswordIdentity(text)))
+
+        case Auth.Key(path0, maybePass) =>
+          Resource liftF {
+            for {
+              path <- blocker.blockOn(F.delay(path0.toAbsolutePath()))
+              provider <- F.delay(new FileKeyPairProvider(path))
+
+              _ <- maybePass match {
+                case Some(password) =>
+                  F delay {
+                    provider setPasswordFinder { (session, key, index) =>
+                      if (key.getName() === path.toString)
+                        password
+                      else
+                        null
+                    }
+                  }
+
+                case None =>
+                  F.delay(provider.setPasswordFinder(FilePasswordProvider.EMPTY))
+              }
+
+              pairs <- blocker.blockOn(F.delay(provider.loadKeys(session)))
+              _ <- pairs.asScala.toList traverse_ { kp =>
+                F.delay(session.addPublicKeyIdentity(kp))
+              }
+            } yield ()
+          }
       }
 
-      r.map(new Process[F](_, blocker, chunkSize))
+      success <- Resource.liftF(fromFuture(F.delay(session.auth())))
+
+      // TODO handle auth failure (minor problems...)
+
+      channel <- Resource.make(
+        for {
+          channel <- F.delay(session.createExecChannel(command))
+          _ <- F.delay(channel.setStreaming(ClientChannel.Streaming.Async))
+          opened <- fromFuture(F.delay(channel.open()))
```

Comment:
Does the command begin executing here? If so, might we miss data from the in/out/err streams? Maybe we can push the `open` into the in/out/err streams on the returned `Process` using an `MVar` as a latch to coordinate?

Reply:
I'd love to hop on a quick call and talk to you about this. There seems to be some major witchcraft here inside of sshd, but whether there is or isn't, what we do about it from *our* side gets really complicated without compromising resource safety.